src/api/cluster_create.go

```diff
@@ -110,9 +110,53 @@ func (a *ClusterAPI) CreateCluster(c *gin.Context) {
 			createClusterRequest.SecretId = secret.GenerateSecretIDFromName(createClusterRequest.SecretName)
 		}
 
-		commonCluster, err := a.createCluster(ctx, &createClusterRequest, orgID, userID, createClusterRequest.PostHooks)
+		existingCluster, err := a.clusterManager.GetClusterByName(ctx, orgID, createClusterRequest.Name)
 		if err != nil {
-			c.JSON(err.Code, err)
+			if err.Error() == "could not get cluster from database: could not find cluster by name: cluster not found" {
+				a.logger.Info("cluster does not exist yet")
+			} else {
+				c.JSON(http.StatusInternalServerError, pkgCommon.ErrorResponse{
+					Code:    http.StatusInternalServerError,
+					Message: "check if the cluster already exists failed",
+					Error:   err.Error(),
+				})
+				return
+			}
+		}
```

Comment:
**Required**: matching exact error messages is very fragile, we should either use error behavior interfaces when provided or at least error type matching when not, because it makes the check more resilient to changes.

For example if we change the error message into "cluster %cluster_name% not found" then this check is going to be broken.

This can be done by utilizing the [Go 1.13 error functions](https://go.dev/blog/go1.13-errors), namely [As](https://pkg.go.dev/emperror.dev/errors#As) (fortunately emperror provides a proxy function for it).

```golang
// src/api/error.go

type notFoundError interface {
		NotFound() bool
}

// isNotFoundError returns true if the error implements the NotFound behavior and it returns true.
func isNotFoundError(err error) bool {
    var err notFoundError

	return errors.As(err, &err) && 
        err.NotFound()
}
```

```golang
// src/api/cluster_create.go

    if err != nil && 
        !isNotFoundError(err) {
        c.JSON(http.StatusInternalServerError, pkgCommon.ErrorResponse{
		    Code:    http.StatusInternalServerError,
		    Message: "check if the cluster already exists failed",
		    Error:   err.Error(),
		})

		return
    } else if err != nil && 
        isNotFoundError(err) {
        a.logger.Info("cluster does not exist yet, creating")
    }
```

Reply:
Thanks for the tip!

Reply:
Updated the example, because I messed up the naming.